### PR TITLE
WELD-2371 Weld SE - allow progr. lookup during AfterDeploymentValidation

### PR DESCRIPTION
**Pull request by mkouba**

### CODE REVIEW

File: environments/se/core/src/main/java/org/jboss/weld/environment/se/WeldContainer.java [hunk 1]
```diff
@@ -238,6 +242,13 @@ private void fireContainerInitializedEvent() {
      */
     public Event<Object> event() {
         checkState();
+        if (event == null) {
+            synchronized (this) {
```
Comment (manovotn) [inline environments/se/core/src/main/java/org/jboss/weld/environment/se/WeldContainer.java]: I am not following here - why do we want this in `sync` block?
Also, why don't we do it just once in sync block in the `complete` method?

File: environments/se/core/src/main/java/org/jboss/weld/environment/se/logging/WeldSELogger.java [hunk 2]
```diff
@@ -73,7 +73,7 @@
 
     // 2013 used in Weld 2.3
 
-    @Message(id = 2014, value = "Weld SE container {0} not initialized completely", format = Format.MESSAGE_FORMAT)
-    IllegalStateException weldContainerNotInitializedCompletely(Object id);
+    @Message(id = 2014, value = "Weld SE container {0} deployment not validated yet - methods for programmatic lookup cannot be used", format = Format.MESSAGE_FORMAT)
```
Comment (manovotn) [inline environments/se/core/src/main/java/org/jboss/weld/environment/se/logging/WeldSELogger.java]: Sounds a bit clumsy, how about a twist:
`Weld SE container with id {0} has not yet validated the deployment - methods ....`

File: environments/se/core/src/main/java/org/jboss/weld/environment/se/WeldContainer.java (hunk 1 above)
Comment (mkouba) [inline environments/se/core/src/main/java/org/jboss/weld/environment/se/WeldContainer.java]: Visibility - `event` is volatile and may be initialized outside the `complete()` method, the same applies to `instance` (that is the point of WELD-2371).

File: environments/se/core/src/main/java/org/jboss/weld/environment/se/logging/WeldSELogger.java (hunk 2 above)
Comment (mkouba) [inline environments/se/core/src/main/java/org/jboss/weld/environment/se/logging/WeldSELogger.java]: Sounds good.

File: environments/se/core/src/main/java/org/jboss/weld/environment/se/WeldContainer.java (hunk 1 above)
Comment (tremes) [inline environments/se/core/src/main/java/org/jboss/weld/environment/se/WeldContainer.java]: So this is basically check whether the Event built-in bean has been already initialized/used right? Still not quite sure why it's volatile. Usage within extension could mean execution in different thread?

Comment (mkouba) [inline environments/se/core/src/main/java/org/jboss/weld/environment/se/WeldContainer.java]: We want to be sure the `event` is only initialized once and visible to all threads, that's the reason why it's volatile and initialized in a synchronized block.

**Pull request merged**